presto-main/src/main/java/com/facebook/presto/operator/SetBuilderOperator.java

```diff
@@ -203,12 +213,35 @@ public void addInput(Page page)
 
         Block sourceBlock = page.getBlock(setChannel);
         Page sourcePage = hashChannel.isPresent() ? new Page(sourceBlock, page.getBlock(hashChannel.get())) : new Page(sourceBlock);
-        channelSetBuilder.addPage(sourcePage);
+
+        // Add the pending work to unfinishedWork
+        unfinishedWork = Optional.of(channelSetBuilder.addPage(sourcePage));
+
+        // Processes the unfinishedWork for this page by adding the data to the hash table. If this page
+        // can't be fully consumed (e.g. rehashing fails), the unfinishedWork will be left with non-empty value.
+        processUnfinishedWork();
     }
 
     @Override
     public Page getOutput()
     {
         return null;
     }
+
+    private boolean processUnfinishedWork()
+    {
+        boolean done = unfinishedWork.get().process();
```

Comment:
add assertion: `checkState(unfinishedWork.isPresent());`

Reply:
Optional.get() would check if it's null. But I'm changing it back to not using Optional, so I'll add it back.